flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -606,6 +627,17 @@ private RocksDB openDB(
 		Preconditions.checkState(1 + stateColumnFamilyDescriptors.size() == stateColumnFamilyHandles.size(),
 			"Not all requested column family handles have been created");
 
+		if (this.metricOptions.isEnabled()) {
+			this.nativeMetricMonitor = new RocksDBNativeMetricMonitor(
+				dbRef,
+				rocksDBResourceGuard,
+				metricOptions,
+				operatorMetricGroup
+			);
+
+			this.cancelStreamRegistry.registerCloseable(nativeMetricMonitor);
```

Comment:
I think this is not a good way for a clean shutdown. Instead I would suggest to close this in the `dispose()` method.

Reply:
Why is this an issue? The cancelStreamRegistry is closed in `AbstractKeyedStateBackend#dispose`.  I manually called close inside of `dispose` first but I was concerned about the strict ordering it imposed because if the nativeMetricsWatcher wasn't closed first `rocksDBResourceGuard.close()` would block forever.

Reply:
It is not an issue in the sense of a bug, but it is much more explicit. `CloseableRegistry` is typically used to coordinate closing of resources across multiple threads and I would not use it if we move operate in just one thread.

I think you need to close metrics as first thing in dispose() or else it will block. As a rule of thump, all closing in `dispose()`should happen in the reverse order of creation.